packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -288,157 +296,124 @@ export class OdspDocumentService implements IDocumentService {
+                );
                 throw connectionError;
-            });
-        }
+            }
+        };
 
-        const startNonAfd = performance.now();
-        return OdspDocumentDeltaConnection.create(
-            tenantId,
-            documentId,
-            token,
-            io,
-            client,
-            url,
-            hasUrl2 ? 15000 : 20000,
-            this.logger,
-        ).then((connection) => {
-            logger.sendTelemetryEvent({ eventName: "UsedNonAfdUrl" });
-            return connection;
-        }).catch(async (connectionError) => {
-            const endNonAfd = performance.now();
-            if (hasUrl2 && this.canRetryOnError(connectionError)) {
-                // eslint-disable-next-line max-len
-                debug(`Socket connection error on non-AFD URL. Error was [${connectionError}]. Retry on AFD URL: ${url2}`);
+        const connectWithAfd = async () => {
+            assert(afdUrl !== undefined, "Tried to connect with AFD but no AFD url provided");
 
-                return OdspDocumentDeltaConnection.create(
+            const startTime = performance.now();
+            try {
+                const connection = await OdspDocumentDeltaConnection.create(
                     tenantId,
                     documentId,
                     token,
                     io,
                     client,
-                    url2!,
+                    afdUrl,
                     20000,
                     this.logger,
-                ).then((connection) => {
-                    // Refresh AFD cache
-                    const cacheResult = this.writeLocalStorage(lastAfdConnectionTimeMsKey, Date.now().toString());
-                    if (cacheResult) {
-                        // eslint-disable-next-line max-len
-                        debug(`Cached AFD connection time. Expiring in ${new Date(Number(localStorage.getItem(lastAfdConnectionTimeMsKey)) + afdUrlConnectExpirationMs)}`);
-                    }
-                    logger.sendPerformanceEvent({
-                        eventName: "UsedAfdUrl",
-                        duration: endNonAfd - startNonAfd,
-                        refreshedCache: cacheResult,
-                        fromCache: false,
-                    }, connectionError);
-
-                    return connection;
-                }).catch((retryError) => {
-                    logger.sendPerformanceEvent({
-                        eventName: "FailedAfdUrlFallback",
-                        duration: endNonAfd - startNonAfd,
-                    }, retryError);
-                    throw retryError;
+                );
+                // Set the successful connection attempt in the cache so we can skip the non-AFD failure the next time
+                // we try to connect and immediately try AFD instead.
+                writeLocalStorage(lastAfdConnectionTimeMsKey, Date.now().toString());
+                const endTime = performance.now();
```

Comment:
I'd move it up one line - measure only network (as much as it's possible) - who knows, maybe writing to localStorage may (?) result in some user prompt

Reply:
Good catch, I missed this one when I was moving the others up.  Fixed in latest